File: app/src/main/java/fuck/wifiadbtoggle/droidvendorssuck/ScheduleAlarmScheduler.java
package fuck.wifiadbtoggle.droidvendorssuck;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.os.Build;

public final class ScheduleAlarmScheduler {
    private ScheduleAlarmScheduler() {
    }

    public static void scheduleNext(Context context) {
        AlarmManager alarmManager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        Intent intent = new Intent(context, ScheduleAlarmReceiver.class);
        PendingIntent pending = PendingIntent.getBroadcast(
            context,
            0,
            intent,
            pendingFlags()
        );

        if (!Settings.isScheduleEnabled(context) || !new ScheduleDbHelper(context).hasAnyEnabled()) {
            alarmManager.cancel(pending);
            return;
        }

        Long next = ScheduleManager.getNextBoundary(context);
        if (next == null) {
            alarmManager.cancel(pending);
            return;
        }
        if (Build.VERSION.SDK_INT >= 19) {
            alarmManager.setExact(AlarmManager.RTC_WAKEUP, next, pending);
        } else {
            alarmManager.set(AlarmManager.RTC_WAKEUP, next, pending);
        }
    }

    private static int pendingFlags() {
        if (Build.VERSION.SDK_INT >= 23) {
            return PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE;
        }
        return PendingIntent.FLAG_UPDATE_CURRENT;
    }
}

File: app/src/main/java/fuck/wifiadbtoggle/droidvendorssuck/ScheduleAlarmScheduler.kt
package fuck.wifiadbtoggle.droidvendorssuck

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.Build

object ScheduleAlarmScheduler {
    fun scheduleNext(context: Context) {
        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val intent = Intent(context, ScheduleAlarmReceiver::class.java)
        val pending = PendingIntent.getBroadcast(
            context,
            0,
            intent,
            pendingFlags()
        )

        val next = ScheduleManager.getNextBoundary(context)
        if (next == null) {
            alarmManager.cancel(pending)
            return
        }
        if (Build.VERSION.SDK_INT >= 19) {
            alarmManager.setExact(AlarmManager.RTC_WAKEUP, next, pending)
        } else {
            alarmManager.set(AlarmManager.RTC_WAKEUP, next, pending)
        }
    }

    private fun pendingFlags(): Int {
        return if (Build.VERSION.SDK_INT >= 23) {
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        } else {
            PendingIntent.FLAG_UPDATE_CURRENT
        }
    }
}
